yt/yt/client/table_client/composite_compare.cpp

```diff
@@ -276,4 +308,112 @@ TFingerprint CompositeFarmHash(TYsonStringBuf value)
 
 ////////////////////////////////////////////////////////////////////////////////
 
+std::optional<TYsonString> TruncateCompositeValue(TYsonStringBuf value, i64 size)
+{
+    YT_ASSERT(value.GetType() == EYsonType::Node);
+
+    if (size <= 0) {
+        return {};
+    }
+
+    TMemoryInput valueIn(value.AsStringBuf());
+    TYsonPullParser valueParser(&valueIn, EYsonType::Node);
+
+    TString truncatedYson;
+    TStringOutput output(truncatedYson);
+    output.Reserve(std::min(size, std::ssize(value.AsStringBuf())));
+    TCheckedInDebugYsonTokenWriter writer(&output);
+
+    i64 unclosedListCount = 0;
+    bool nonEmptyResult = false;
+
+    for (auto remainingBytes = size;;) {
+        const auto item = valueParser.Next();
+
+        // Both braces of the outmost list are not considered to be inside an enclosing list.
+        bool isInsideList = unclosedListCount - (item.GetType() == EYsonItemType::EndList);
```

Comment:
Subtracting bool from int and storing into bool seems a bit scary to me. Let's rewrite it in a more explicit way.

Reply:
Rewrote it into a more concise condition.